File: TwoPointers/lc15/ThreeSum.py
class Solution(object):
    def threeSum(self, nums):
        """
        :type nums: List[int]
        :rtype: List[List[int]]
        """

        #异常处理
        if not nums:
            return []
        
        #答案集建立和排序
        result = []
        nums.sort()

        #3sum 实质就是一个for循环的遍历 + 2sum。把a + b + c = 0 看成 a + b= -c，for循环-c套一个a + b的2sum。
        for a in range(0, len(nums) - 2):
            left, right = a + 1, len(nums) - 1
            while left < right: 
                if nums[left] + nums[right] == -nums[a]:
                    #去重
                    if [nums[a], nums[left], nums[right]] not in result:
                        result.append([nums[a], nums[left], nums[right]])
                    left += 1
                    right -= 1
                elif nums[left] + nums[right] < -nums[a]:
                    left += 1
                else:
                    right -= 1 
        
        return result

File: TwoPointers/lc15/ThreeSum.java
package TwoPointers.lc15;

import java.util.*;

public class ThreeSum {
  public static List<List<Integer>> search(int[] arr) {
    List<List<Integer>> triplets = new ArrayList<>();
    //先排序
    Arrays.sort(arr);
    for (int i = 0; i < arr.length - 2; i++) {
      //因为按升序排所以只要有大于0的就一定不可能找到sum == 0的组合
      if (arr[i] > 0) {
        break;
      }
      //去除重复
      if (i > 0 && arr[i] == arr[i - 1]) {
        continue;
      }
      int start = i + 1;
      int end = arr.length - 1;
      while (start < end) {
        int sum = arr[i] + arr[start] + arr[end];
        if (sum == 0) {
          triplets.add(Arrays.asList(arr[i], arr[start], arr[end]));
          start++;
          end--;
          //去除重复
          while (start < end && arr[start] == arr[start - 1]) {
            start++;
          }
          //去除重复
          while (start < end && arr[end] == arr[end + 1]) {
            end--;
          }
        } else if (sum < 0) {
          start++;
        } else {
          end--;
        }
      }
    }
    return triplets;
  }
}
